simtools/util/input_validation.py

```diff
@@ -0,0 +1,114 @@
+import datetime
+import logging
+import re
+
+import simtools.util.general as gen
+
+class SchemaValidator:
+    """
+    Validate a dictionary against a reference schema
+
+    """
+
+    def __init__(self, reference_schema_file, data_dict):
+        """
+        Initalize validation class and read required
+        reference schema
+
+        Parameters:
+        -----------
+        reference_schema: dict
+            reference schema
+        data_dict: dict
+            dictionary to be validated
+
+        """
+
+        self._logger = logging.getLogger(__name__)
+
+        if reference_schema_file:
+            self._reference_schema = gen.collectDataFromYamlOrDict(
+                reference_schema_file, None)
+        self.data_dict = data_dict
+
+    def validate(self):
+        """
+        Schema validation
+
+        """
+        self._validate_schema(
+            self._reference_schema,
+            self.data_dict)
+
+    def _validate_data_type(self, schema, key, data_field):
+        """
+        Validate data type against the expected data type
+        from schema
+
+        Raises value error if data types are inconsistent
+        """
+        self._logger.debug('checking data field {} for {}'.format(
+            key, schema['type']))
+
+        if schema['type'] == 'datetime':
+            format = "%Y-%m-%d %H:%M:%S"
+            try:
+                datetime.datetime.strptime(data_field, format)
+            except:
+                raise ValueError(
+                    'invalid date format. Expected {}; Found {}'.format(
+                        format, data_field))
+
+        elif schema['type'] == 'email':
+            regex = r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b'
+            if not re.fullmatch(regex, data_field):
+                raise ValueError(
+                    'invalid email format in field {}: {}'.format(
+                        key, data_field))
+
+        elif type(data_field).__name__ != schema['type']:
```

Comment:
We could be a little more forgiving here and check first if we can convert to the correct type. It happens that a number is read as a string in yaml and a simple `float(entry)` works. I am not sure we should be forgiving, it opens up possible mistakes, but maybe you want to take it into account.

Reply:
It won't be pretty, but we can define a converter dict like 

`convert = {'str': type('str'), 'float': type(1.0), 'int': type(0), 'bool': type(True)}`

and use a `try` and `except` like in the code below

```
elif type(data_field).__name__ != schema['type']:
    try:
        _value = convert[schema['type']](data_field)
    except ValueError:
        self._logger.error('invalid data type for key {}. Expected: {}, Found: {}'.format(
                    key, schema['type'], type(data_field).__name__))
        raise
```

I don't think it's necessarily good. Just a thought if we want to be forgiving.

Reply:
Implemented this.

But not entirely convinced: while converting int to floats is fine, I am hesitating about the opposite: converting 2.3 to 2 works, but maybe it is not what we want: when users give a value as a float, we should note convert it to an integer.

Reply:
Absolutely we shouldn't. I only meant converting from string to float, int or boolean, nothing else. Should have written that explicitly.

Reply:
I do think we need to be explicit and allow conversion from string or int, but not from e.g. float to int. I am allowing a bit of forgiveness now only for string and int.

Reply:
OK!